### ISSUE TITLE
Chart.yaml's kubeVersion field on GKE - unusable?

### ISSUE THREAD

**Issue by consideRatio** [kubernetes/helm#3810]
### CLOSING UPDATE
In order to allow pre-releases such as googles `v1.9.4-gke.1` the constraint `kubeVersion: ">=1.8.0"` must be changed to `kubeVersion: ">=1.8.0-something"` for example `kubeVersion: ">=1.8.0-0"`.

### Situation
I made a Helm upgrade (with v2.8.2 client/server) on my GKE cluster on a Chart.yaml with `kubeVersion: ">=1.8.0"` and received the following error.

```
Error: UPGRADE FAILED: Chart requires kubernetesVersion: >=1.8.0 which is incompatible with Kubernetes v1.9.4-gke.1
```

### For reference
* [This is an online SemVer checker](http://jubianchi.github.io/semver-check/)
* The error is throwed here...
https://github.com/kubernetes/helm/blob/8fef5aad52ad2d7aeb32cbccd73f54623717fa6f/pkg/tiller/release_server.go#L271-L278
* Because the behavior of the IsCompatibleRange function...
https://github.com/kubernetes/helm/blob/58e4b6ac61938fe056037826225dd19562c0ed9f/pkg/version/compatible.go#L48-L61

### Question
What should be done? With my current understanding of the situation, the usage of `kubeVersion` is contraproductive if Google Cloud is involved. Perhaps this is an issue for them rather than you at helm?

**Comment by unguiculus** [comment]
Related: https://github.com/kubernetes/charts/issues/3002

cc @mattfarina

**Comment by michelleN** [comment]
Hey @consideRatio - thanks for bringing this up! We've seen this type of issue before. Semver (as you probably already saw) denotes anything with a hyphen and something after it as a pre-release which doesn't guarantee stability and therefore doesn't satisfy the constraint. 

I wonder why gke is not using a `+gke.1` instead of `-gke.1`. Is GKE running a version of Kubernetes that is not compatible with upstream Kubernetes? cc/ @kow3ns  

Nonetheless, after discussing with other core maintainers it seems like there may be a work around involving an `-r0` at the end of the kubeVersion in our logic. I tried to get that to work today but couldn't. cc/ @mattfarina @prydonius @bacongobbler

**Comment by consideRatio** [comment]
I asked [masterminds/semver](https://github.com/Masterminds/semver/issues/69) about this indirectly.

**Comment by consideRatio** [comment]
@michelleN apparently it is intentional behavior that was not at all obvious to me, that the constraint ">=1.8.0" should exclude "1.9.4-gke.1" because...

> SemVer comparisons without a pre-release value will skip pre-release versions. For example, >1.2.3 will skip pre-releases when looking at a list of values while >1.2.3-alpha.1 will evaluate pre-releases. - [Source](https://github.com/Masterminds/semver#basic-comparisons)

Confusion remains for me as to why ">0" allows for example "0.0.1-alpha" while ">0.0" doesn't for the reason above, but I'll put this to rest now.

This issue is now perhaps a documentation issue rather than a bug issue. Close?

**Comment by mattfarina** [comment]
Interesting timing on this. [I opened an issue with GKE about this topic last week](https://issuetracker.google.com/issues/77503699).

There is a simple trick to cause the semver resolver to evaluate pre-releases and that's to do something like: `>=1.8.0-0`.

The reasons for this are two fold.

1. When the resolver sees a pre-release in a comparison range it will them evaluate them (opt-in model)
1. According to the [semver spec](https://semver.org/#spec-item-9) a pre-release (the part after the -) can only be the ASCII characters of `[0-9A-Za-z-]`. The lowest [ASCII character](http://www.asciitable.com/), when sorting is 0.

The reason pre-releases are skipped by default has to do with the spec noting, "A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version." In the past the resolved didn't skip them and it turned out to be a practical problem.

There are two things I'd really like...

* GKE to not break from the SemVer spec. Some tools don't include pre-releases at all in comparisons. Other libs, like [node-semver](https://github.com/npm/node-semver), perform the same way Helm does and only includes them if a comparator has a pre-release on it.
* I wish it were easier for Masterminds/semver to toggle on and off the pre-release programming programmatically. Last time I looked at that is was hard to do without breaking the v1 API. We'll see how v2 ends up landing on that. It just wasn't an easy win with the v1 API. Though, if someone wants to try I'm open to it.

Some more details (and the masterminds/semver project may need some better docs or even a bug fix)... `0` and `0.0` are really `0.0.0` under the hood. This is a special case situation to make sure that some of the strange comparisons around `0.0.0` happen correctly. I'll try to revisit that with some more tests.

**Comment by mattfarina** [comment]
@consideRatio could you take a look at https://github.com/Masterminds/semver/pull/70

**Comment by mattfarina** [comment]
@consideRatio if you would also look at https://github.com/Masterminds/semver/pull/71 which adds more docs on pre-releases I'd appreciate it

**Comment by consideRatio** [comment]
Closing this issue now! Thanks for your help with this!

For reference, fixes @mattfarina made was released in [Masterminds/semver v1.4.2](https://github.com/Masterminds/semver/releases/tag/v1.4.2).
